lib/NeuraDialect/Transforms/CanonicalizeLiveInPass.cpp

```diff
@@ -17,24 +21,275 @@ using namespace mlir;
 #include "NeuraDialect/NeuraPasses.h.inc"
 
 namespace {
-LogicalResult promoteLiveInValuesToBlockArgs(Region &region) {
+struct DirectDataflowLiveIn {
+  // The live-in value.
+  Value value;
+  // The block where the live-in value is defined.
+  Block *defining_block;
+  // The block where the live-in value is used.
+  Block *using_block;
+};
+
```

Comment:
Add comment to explain this func, rename func to `isDominatingEachOther()` or some other better name.
```
//   A
// B   C (draw some arrows)
//   D
//
// A is D's domainator and D is A's post-dominator,
// and A -> D is through `cond_br`.
```